### PR TITLE
Refactor recommendation code 

### LINKED ISSUE #1101: Recommendation Code Cleanup
The recommendation code structure we have currently has a lot of redundancy. Also, the Term and Performance Profile classes aren't properly divided into their respective packages.

We need to:

-  Refactor the classes and the packages.
-  Come up with a Design to segregate the classes into the correct packages
-  Add new sub-package for the Term data
- Add or remove existing abstraction wherever required.

### PR DESCRIPTION
**Pull request by khansaad**
This PR fixes #1101

### CODE REVIEW

File: src/main/java/com/autotune/analyzer/services/UpdateRecommendations.java [hunk 1]
```diff
@@ -102,111 +90,16 @@ protected void doPost(HttpServletRequest request, HttpServletResponse response)
...
-                    e.printStackTrace();
-                    LOGGER.error("Failed to create recommendation for experiment: {} and interval_start_time: {} and interval_end_time: {}",
-                            experiment_name,
-                            interval_start_time,
-                            interval_end_time);
-                    sendErrorResponse(response, null, HttpServletResponse.SC_INTERNAL_SERVER_ERROR, e.getMessage());
-                    return;
-                }
+            // validate the input params and generate recommendations
+            RecommendationEngine recommendationEngine = new RecommendationEngine();
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/analyzer/services/UpdateRecommendations.java:94]: Lets not call empty constructor
initialize RecommendationEngine(experiment_name , interval_endtime)

File: src/main/java/com/autotune/analyzer/services/UpdateRecommendations.java [hunk 2]
```diff
@@ -102,111 +90,16 @@ protected void doPost(HttpServletRequest request, HttpServletResponse response)
...
+            KruizeObject kruizeObject = recommendationEngine.validateAndGenerateRecommendation(experiment_name,
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/analyzer/services/UpdateRecommendations.java:95]: have separate function
Validate()
generateRecommednations()

and create  getKruizeObject()

File: src/main/java/com/autotune/analyzer/services/UpdateRecommendations.java [hunk 3]
```diff
@@ -102,111 +90,16 @@ protected void doPost(HttpServletRequest request, HttpServletResponse response)
...
+                    interval_end_time, interval_start_time, intervalEndTimeStr, intervalStartTimeStr, calCount);
+            if (!kruizeObject.getValidation_data().isSuccess()) {
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/analyzer/services/UpdateRecommendations.java:97]: Let kruizeObject be POJO and remove getValidation_data() function from this

File: src/main/java/com/autotune/analyzer/services/UpdateRecommendations.java (hunk 2 above)
Comment (msvinaykumar) [reply src/main/java/com/autotune/analyzer/services/UpdateRecommendations.java:95]: and lets be empty function call , lets not pass params

File: src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java [hunk 4]
```diff
@@ -1,394 +1,625 @@
...
+import static com.autotune.analyzer.recommendations.RecommendationConstants.RecommendationValueConstants.MEM_ZERO;
+import static com.autotune.analyzer.utils.AnalyzerErrorConstants.AutotuneObjectErrors.MISSING_EXPERIMENT_NAME;
+
+public class RecommendationEngine {
+    private String performanceProfile;
+    private String experimentName;
+    private Timestamp intervalEndTime;
+    private Map<String, Terms> terms;
+    List<RecommendationModel> recommendationModels;
+
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java:50]: create KruizeObject parameter with get and set

File: src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java [hunk 5]
```diff
@@ -1,394 +1,625 @@
...
+    private static final Logger LOGGER = LoggerFactory.getLogger(RecommendationEngine.class);
+
+
+    public RecommendationEngine(String performanceProfile, String experimentName, Timestamp intervalEndTime, Map<String, Terms> terms) {
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java:54]: RecommendationEngine should take only two params 
experiment_name and interval_end_time 

remaining parameters should be derived from KruizeObject

File: src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java [hunk 6]
```diff
@@ -1,394 +1,625 @@
...
-     */
-    private static int getNumPods(Map<Timestamp, IntervalResults> filteredResultsMap) {
-        Double max_pods_cpu = filteredResultsMap.values()
-                .stream()
-                .map(e -> {
-                    Optional<MetricResults> cpuUsageResults = Optional.ofNullable(e.getMetricResultsMap().get(AnalyzerConstants.MetricName.cpuUsage));
-                    double cpuUsageSum = cpuUsageResults.map(m -> m.getAggregationInfoResult().getSum()).orElse(0.0);
-                    double cpuUsageAvg = cpuUsageResults.map(m -> m.getAggregationInfoResult().getAvg()).orElse(0.0);
-                    double numPods = 0;
+    private void init() {
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java:66]: init() 
inside init() methode call validation 
and if validation is sucess
Create KruizeObject 
and derive other parameters like
        datasource , metricProfile , Terms etc

File: src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java [hunk 7]
```diff
@@ -1,394 +1,625 @@
...
-                        cpuRequestInterval = cpuUsageTotal;
+        interval_end_time = Utils.DateUtils.getTimeStampFrom(KruizeConstants.DateFormats.STANDARD_JSON_DATE_FORMAT,
+                intervalEndTimeStr);
+        LOGGER.debug("UpdateRecommendations API request count: {} experiment_name : {} and interval_start_time : {} and interval_end_time : {} ", calCount, experiment_name, intervalStartTimeStr, intervalEndTimeStr);
+        LOGGER.debug("experiment_name : {} and interval_start_time : {} and interval_end_time : {} ", experiment_name, intervalStartTimeStr, intervalEndTimeStr);
+
+        Map<String, KruizeObject> mainKruizeExperimentMAP = new ConcurrentHashMap<>();
+        Map<String, Terms> terms = new HashMap<>();
+        try {
+            new ExperimentDBService().loadExperimentFromDBByName(mainKruizeExperimentMAP, experiment_name);
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java:146]: abstract out for getting data
from DataSource identify where to fetch data  is it via DB or RestAPI

File: src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java [hunk 8]
```diff
@@ -1,394 +1,625 @@
...
+                // set the default terms if the terms object is empty
+                if (kruizeObject.getTerms() == null)
+                    setDefaultTerms(terms);
+                // get the datasource
+                String dataSource = kruizeObject.getDataSource(); // TODO: This will be required for future use
+                int maxDay = Terms.getMaxDays(terms);
+                interval_start_time = Timestamp.valueOf(Objects.requireNonNull(interval_end_time).toLocalDateTime().minusDays(maxDay));
+
+                mainKruizeExperimentMAP.put(experiment_name, kruizeObject);
+                new ExperimentDBService().loadResultsFromDBByName(mainKruizeExperimentMAP, experiment_name,
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java:160]: abstract out for getting data
from DataSource identify where to fetch data is it via DB or RestAPI

File: src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java (hunk 7 above)
Comment (msvinaykumar) [reply src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java:146]: Ok getting KruizeObject is always from DB not via restAPI
But before fetching results always check for is it via DB or via restAPI

File: src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java [hunk 9]
```diff
@@ -1,394 +1,625 @@
...
+                        interval_start_time, interval_end_time);
+
+                // generate recommendation
+                try {
+                    generateRecommendations(kruizeObject);
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java:165]: Have abstraction for
Terms and Models
it should take two parameter container Name as a parameter and  <ts,IntervalResults> results
and return recommendations 
should be set Container

File: src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java (hunk 8 above)
Comment (khansaad) [reply src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java]: Done

File: src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java (hunk 7 above)
Comment (khansaad) [reply src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java]: Done

File: src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java (hunk 6 above)
Comment (khansaad) [reply src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java:61]: Done

File: src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java (hunk 5 above)
Comment (khansaad) [reply src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java]: Done

File: src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java (hunk 4 above)
Comment (khansaad) [reply src/main/java/com/autotune/analyzer/recommendations/engine/RecommendationEngine.java:50]: Done

**Comment by khansaad** [comment]
Test results attached after running on jenkins.
[kruize_functional_test_results_pr-1105.zip](https://github.com/kruize/autotune/files/14667627/kruize_functional_test_results_pr-1105.zip)

File: src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfile.java [hunk 10]
```diff
@@ -36,6 +39,8 @@ public class PerformanceProfile {
     @SerializedName("slo")
     private SloInfo sloInfo;
 
+    private Map<String, Terms> terms;
```
Comment (msvinaykumar) [inline src/main/java/com/autotune/analyzer/performanceProfiles/PerformanceProfile.java:42]: Please double-check whether the perfprofile is associated with Terms. If it's not related, then it doesn't make sense to have a reference in the perf profile.

**Review by dinogun — APPROVED** [review]
Approving this based on all functionality tests passing.

**Pull request merged**